File: codes/statisticalAnalysisLearning/statisticalLearning/FirstCourseMachineLearningCode/matlab/chapter5/knnexample.m
%% knnexample.m
% From A First Course in Machine Learning, Chapter 5.
% Simon Rogers, 01/11/11 [simon.rogers@glasgow.ac.uk]
% Setting K in KNN
clear all;close all;

%% Generate some data
N1 = 100; N2 = 30; % Class sizes
x = [randn(N1,2);randn(N2,2)+3];
t = [repmat(0,N1,1);repmat(1,N2,1)];
N = size(x,1);
%% Plot the data
ma = {'ko','ks'};
fc = {[0 0 0],[1 1 1]};
tv = unique(t);
figure(1); hold off
for i = 1:length(tv)
    pos = find(t==tv(i));
    plot(x(pos,1),x(pos,2),ma{i},'markerfacecolor',fc{i});
    hold on
end

%% Generate the decision boundaries for various values of K
[Xv Yv] = meshgrid(min(x(:,1)):0.1:max(x(:,1)),min(x(:,2)):0.1:max(x(:,2)));
% Loop over test points

Kvals = [1 2 5 10 20 50 59];
for kv = 1:length(Kvals)
    %%
    classes = zeros(size(Xv));
    K = Kvals(kv);
    for i = 1:length(Xv(:))
        this = [Xv(i) Yv(i)];
        dists = sum((x - repmat(this,N,1)).^2,2);
        [d I] = sort(dists,'ascend');
        [a,b] = hist(t(I(1:K)));
        pos = find(a==max(a));
        if length(pos)>1
            order = randperm(length(pos));
            pos = pos(order(1));
        end
        classes(i) = b(pos);
    end
    figure(1); hold off
    for i = 1:length(tv)
        pos = find(t==tv(i));
        plot(x(pos,1),x(pos,2),ma{i},'markerfacecolor',fc{i});
        hold on
    end
    contour(Xv,Yv,classes,[0.5 0.5],'k')
    ti = sprintf('K = %g',K);
    title(ti);
end

File: codes/statisticalAnalysisLearning/statisticalLearning/FirstCourseMachineLearningCode/R/chapter5/knnexample.R
## knnexample.R
# From A First Course in Machine Learning, Chapter 4.
# Francois-Xavier Briol, 04/06/16 [f-x.briol@warwick.ac.uk]
# Setting K in KNN
rm(list=ls(all=TRUE))

## Generate some data
N1 = 100
N2 = 30 # Class sizes
x = rbind(matrix(rnorm(N1*2),ncol=2),matrix(rnorm(N2*2)+3,ncol=2))
t = matrix(c(rep(0,N1),rep(1,N2)),ncol=1)
N = nrow(x)

## Plot the data
tv = as.vector(unique(t))
ma = c(0,16)
col= c("blue","red")
pos = which(t==tv[1])
plot(x[pos,1],x[pos,2],col=col[1],pch=ma[1],xlim=c(-3,6),ylim=c(-3,6))
for(i in 1:length(tv)){
  pos = which(t==tv[i])
  points(x[pos,1],x[pos,2],col=col[i],pch=ma[i])
}

## Generate the decision boundaries for various values of K
gridvals = seq(from=min(x),to=max(x),by=0.2)
mesh = meshgrid(gridvals,gridvals)
Xv = mesh$X; Yv=mesh$Y

# Loop over test points
Kvals = c(1,2,5,10,20,50,59)
for(kv in 1:length(Kvals)){
  ##
  classes = matrix(0,ncol=dim(Xv)[1],nrow=dim(Xv)[2])
  K = Kvals[kv]
  for(i in 1:nrow(Xv)){
    for(j in 1:ncol(Xv)){
      this = c(Xv[i,j],Yv[i,j])
      dists = rowSums((x - matrix(rep(this,N),ncol=2))^2)
      d = sort(dists)
      I = order(dists)
      tab = table(t[I[1:K],])
      a = as.vector(tab)
      b = as.numeric(names(tab)) 
      pos = which(a==max(a))
      if(length(pos)>1){
        ordering = sample.int(length(pos))
        pos = pos[ordering[1]]
      }
      classes[i,j] = b[pos]
    }
  }
  pos = which(t==tv[1])
  maintitle = paste("K=",K)
  plot(x[pos,1],x[pos,2],col=col[1],pch=ma[1],xlim=c(min(x[,1]),max(x[,2])),
       ylim=c(min(x[,2]),max(x[,2])),main=maintitle)
  for(i in 1:length(tv)){
    pos = which(t==tv[i])
    points(x[pos,1],x[pos,2],col=col[i],pch=ma[i])
  }
  contour(gridvals,gridvals,classes,add=TRUE)
}








